fep/000-manipulating-kubernetes-objects.md

```diff
@@ -0,0 +1,132 @@
+
+## Summary
+
+### Goals
+
+Provide an opinionated CRUD framework for Kubernetes' object manipulation.
+
+### Non-goals
+
+Provide direct access to Kubernetes' REST API.
+
+## Manipulations
+
+Let us start with enumerating operationTypes that this framework shall support.
+These will implement all common object manipulation patterns.
+
+### Creating
+
+* `CreateorUpdate` — accept a Kubernetes object.
+  It retrieves an object, and if it already exists computes a JSON Merge Patch and applies it (will not update .status field).
+  If it does not exist, we create the object.
+
+Optionally, we might support the following Manipulations:
+
+* `Create` — will fail if an object already exists
+* `CreateIfNotExists` — create and object if such an object does not already 
+  exist by namespace/name.
+
+These manipulations are potentially dangerous and useless, since they do not reconcile an object in any way.
+
+### Deleting
+
+Mirrors all REST DELETE [options](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection).
+
+* `EnsureDeleted`
+* `EnsureDeleteBackground` 
+* `EnsureDeleteOrphan`
+
+### Modification (each supports optional status subresource)
+
+#### Considerations
+
+1. **Bandwidth**. Raw `JSONPatch` and `MergePatch` operations will be used when performance of a hook is of utmost importance.
+2. **Field ownership and locking**.
+   * *Optimistic* operationTypes work well for ensuring object's consistency. An object 
+        will be considered as a whole. If there is a conflict, hook's logic will re-run to ensure
+        that object will transform to the desired state
+   * *Fire and forget*. `JQPatch` and `FilterPatch` will work great with ensuring consistency of the whole object, but may miss
+        some updates to an object if there were `resourceVersion` conflicts.
+
+#### Implementations
+
+##### Optimistic locking
+
+These operationTypes will utilize REST's PUT verb. If an operation fails due to a `resourceVersion` conflict,
+shell-operator shall execute a hook again, but not count it as a failure. It should increment an appropriate metric counter
+(`{PREFIX}hook_run_conflicts_total`) which will signal the User about a potential bottleneck.
+
+* `JQPatchOptimistic` — receives a JQFilter that describes object transformations. Only used for shell hooks.
+* `FilterPatchOptimistic` — receives an anonymous function that describes object transformations. Only used for Go hooks.
+
+##### Fire and forget
+
+These work similar to their Optimistic-locking counterparts with one crucial difference. If there is a `resourceVersion`
+conflict, shell-operator shall not call the hook again. Instead, it will apply jq filter or transformations in the anonymous Go function,
+until resource update is successful. With [an exponential backoff](https://pkg.go.dev/k8s.io/client-go@v0.19.4/util/retry#RetryOnConflict)
+and upper limit, of course.
+
+* `JQPatch`
+* `FilterPatch`
```

Comment:
As I see it, JQPatch is a special case of a FilterPatch. We will generate an anonymous function based on a jq-expression and use it internally. Hooks that have written in Go will be able to set this anonymous function directly. 

I am a little bit puzzled because internal concepts and user interface mixed :)

Reply:
That is true. Let's see how I can rephrase this...

Reply:
Go hooks and shell hooks are separate right now.